client/src/main/java/io/pravega/client/segment/impl/AsyncSegmentEventReaderImpl.java

```diff
@@ -0,0 +1,380 @@
+         * @param offset the offset to prepare for.
+         */
+        public CompletableFuture<ByteBuffer> reset(long offset) {
+            if (!this.outstandingPromise.isDone()) {
+                this.outstandingPromise.cancel(false);
+            }
+            this.outstandingPromise = new CompletableFuture<>();
+            this.headerReadingBuffer.clear();
+            this.resultBuffer = null;
+            this.offset = offset;
+            return this.outstandingPromise;
+        }
+
+        /**
+         * Updates the read state based on the given {@link SegmentRead}.
+         * Note that the {@code segmentRead} is progressively consumed by successive event reads.
+         *
+         * @param segmentRead a read response.
+         */
+        public void update(SegmentRead segmentRead) {
+            checkState(!isSuccessful());
+            verifyIsAtCorrectOffset(segmentRead);
+
+            if (resultBuffer == null) {
+                // consume header data from the SegmentRead data buffer
+                if (headerReadingBuffer.hasRemaining()) {
+                    assert headerReadingBuffer.limit() == WireCommands.TYPE_PLUS_LENGTH_SIZE;
+                    offset += fill(headerReadingBuffer, segmentRead.getData());
+                }
+
+                if (!headerReadingBuffer.hasRemaining()) {
+                    // header is ready
+                    headerReadingBuffer.flip();
+                    int type = headerReadingBuffer.getInt();
+                    int length = headerReadingBuffer.getInt();
+                    if (type != WireCommandType.EVENT.getCode()) {
+                        throw new InvalidMessageException("Event was of wrong type: " + type);
+                    }
+                    if (length < 0 || length > WireCommands.MAX_WIRECOMMAND_SIZE) {
+                        throw new InvalidMessageException("Event of invalid length: " + length);
+                    }
+                    resultBuffer = ByteBuffer.allocate(length);
+                }
+            }
+
+            if (resultBuffer != null) {
+                // consume result data from the SegmentRead data buffer
+                if (resultBuffer.hasRemaining()) {
+                    offset += fill(resultBuffer, segmentRead.getData());
+                }
+
+                if (!resultBuffer.hasRemaining()) {
+                    // result is ready
+                    resultBuffer.flip();
+                    complete();
+                }
+            }
+
+            if (!isDone() && segmentRead.isEndOfSegment()) {
+                // unable to fulfill due to end-of-segment
+                assert segmentRead.getData().remaining() == 0;
+                endOfSegment();
+            }
+        }
+
+        private void verifyIsAtCorrectOffset(SegmentRead segmentRead) {
+            checkState(isAtCorrectOffset(segmentRead), "ReadSegment returned data for the wrong offset %s vs %s",
```

Comment:
I'm not sure the correct thing to do in response to a late reply from the server is throw an exception that will kill the connection.

Reply:
This is consistent with the earlier implementation, and won't happen in practice because we cancel any old request before starting a new one.